### PR TITLE
[WIP]Issue 265 move metrics under experienceevent contexts

### PR DESCRIPTION
**Pull request by cdegroot-adobe**
Completed via other PRs that broke down the changes. 

UPDATE: This is being broken up into smaller more manageable chunks.

PR to update ExperienceEvent so that metrics are under the contexts that they relate to.

Relates to issue #265

### CODE REVIEW

**Comment by cdegroot-adobe** [comment]
@kstreeter this needs to be in an interim 0.9.1 for ADC. Can you please review asap.

**Comment by cdegroot-adobe** [comment]
@cluby-adobe, can you please review this. Thanks C.

File: schemas/channels/application.schema.json [hunk 1]
```diff
@@ -1,41 +1,96 @@
...
+    "Applications that are installed on desktop or laptop systems.",
+  "meta:extensible": true,
+  "meta:extends": [
+    "https://ns.adobe.com/xdm/channels/channel#/definitions/channel"
+  ],
+  "type": "object",
+  "meta:abstract": false,
   "definitions": {
-    "application": {
+    "channel": {
```
Comment (cluby-adobe) [inline schemas/channels/application.schema.json]: This channel stuff seems unrelated to the metrics move.  I would suggest removing these changes from this PR to keep the change clean.

File: schemas/context/application.schema.json [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  "$id": "https://ns.adobe.com/xdm/context/application",
+  "$schema": "http://json-schema.org/draft-06/schema#",
+  "title": "Application",
+  "type": "object",
+  "description":
+    "The entites related to Application usage.",
+  "definitions": {
+    "application": {
+      "properties": {
+        "@id": {
```
Comment (cluby-adobe) [inline schemas/context/application.schema.json]: Did the app scheme not exist at all?  Is is this in the PR because of some other change?  Are you reorganizing application along with moving the metrics around?

File: schemas/context/webinfo.schema.json [hunk 3]
```diff
@@ -30,6 +30,14 @@
           "$ref": "https://ns.adobe.com/xdm/context/webreferrer",
           "description":
             "The referrer of a web interaction, which is the URL a visitor came from immediately before the current web interaction was recorded."
+        },
+        "https://ns.adobe.com/xdm/data/metrics/web/link-clicks": {
+          "$ref": "https://ns.adobe.com/xdm/data/measure",
+          "description": "Click of a web-link has occurred."
+        },
+        "https://ns.adobe.com/xdm/data/metrics/web/page-views": {
```
Comment (cluby-adobe) [inline schemas/context/webinfo.schema.json]: I would put this under `xdm:webPageView`

File: schemas/context/webinfo.schema.json [hunk 4]
```diff
@@ -30,6 +30,14 @@
           "$ref": "https://ns.adobe.com/xdm/context/webreferrer",
           "description":
             "The referrer of a web interaction, which is the URL a visitor came from immediately before the current web interaction was recorded."
+        },
+        "https://ns.adobe.com/xdm/data/metrics/web/link-clicks": {
```
Comment (cluby-adobe) [inline schemas/context/webinfo.schema.json]: I would put this under `xdm:webInteraction`

File: schemas/context/application.schema.json (hunk 2 above)
Comment (cdegroot-adobe) [reply schemas/context/application.schema.json]: It was already in the schema, but was strangely referenced through channels. I refactored t to be consistent with the other structures (commerce etc). I also then had to clean up the \channels\ in how it referred to applications. So really just refactoring, nothing net new.

File: schemas/context/webinfo.schema.json (hunk 3 above)
Comment (cdegroot-adobe) [reply schemas/context/webinfo.schema.json]: I am concerned about littering metrics in multiple levels of the hierarchy if we can get around it. I would love more peoples opinions on this.

File: schemas/context/webinfo.schema.json (hunk 4 above)
Comment (cdegroot-adobe) [reply schemas/context/webinfo.schema.json]: As comment above. are we sure we are ok with having metrics in multiple levels of the hierarchy?

File: schemas/channels/application.schema.json (hunk 1 above)
Comment (cdegroot-adobe) [reply schemas/channels/application.schema.json]: The refactoring links this change in to this PR unfortunately. I don't thing I can break it down much more.

@cmathis can we have @jwen-adobe test this branch please and provide feedback?

Comment (jwen-adobe) [reply schemas/channels/application.schema.json]: @cdegroot-adobe The meta:enum value you added "https://ns.adobe.com/xdm/channel-types/application": "Desktop Applications" caused schema merging conflict error during the conversion. I think we need to add this enum value to the @type field of all the other channel related schemas to make them consistent??

File: schemas/context/webinfo.schema.json (hunk 3 above)
Comment (cluby-adobe) [reply schemas/context/webinfo.schema.json]: The benefit is putting the metric right next to the dimension you are most likely looking for to break it down.

Comment (cluby-adobe) [reply schemas/context/webinfo.schema.json]: @cdegroot-adobe, Can we move this under `xdm:webPageView`?

File: schemas/context/webinfo.schema.json (hunk 4 above)
Comment (cluby-adobe) [reply schemas/context/webinfo.schema.json]: Can we move this under `xdm:webInteraction`?

File: schemas/channels/application.schema.json (hunk 1 above)
Comment (jwen-adobe) [reply schemas/channels/application.schema.json]: A preview version of the converted schema is here: https://git.corp.adobe.com/experience-platform/sdm-specification/tree/issue265/json/release

File: schemas/context/webinfo.schema.json (hunk 3 above)
Comment (cdegroot-adobe) [reply schemas/context/webinfo.schema.json]: done

File: schemas/context/webinfo.schema.json (hunk 4 above)
Comment (cdegroot-adobe) [reply schemas/context/webinfo.schema.json]: done

**Comment by cluby-adobe** [comment]
@jwen-adobe - Can you produce a new XED conversion of this branch including the Protobuf with @cdegroot-adobe's latest changes for one final review?

**Comment by lrosenthol** [comment]
There seem like a *lot* of breaking changes going on here.  Is there some background on this?  For example, this discussion of moving metrics inside of other things…

**Comment by cluby-adobe** [comment]
There is also the user experience concern.  Having all of the XDM metrics under a single metrics space causes inconsistency with the layout and discoverability of other custom or solution specific metrics and also caters to a use-case that is not that common.  The more common use-case will be customers wanting to look at the metrics and related dimensions together so it makes sense to lay them out together to help with easier discovery.

**Comment by cdegroot-adobe** [comment]
All, considering #284 there is likely to be major simplification of the URIs for the metric namespaces.

**Comment by cdegroot-adobe** [comment]
@kstreeter - I would like a final decision on this PR please as it touches a lot of files. If we can approve #284 before hand then I can roll those namespace simplifications in here and we can close this.

Note: this PR seems very large, but it is really doing a lot of simple changes over a lot of schemas. We will have feedback shortly on how this PR is working with Analytics at they got a drop of the protobuf lib based on this PR.

**Comment by trieloff** [comment]
@cdegroot-adobe is this still WIP? Can you resolve the conflicts?

**Comment by cdegroot-adobe** [comment]
Hi @trieloff. This is being broken up "UPDATE: This is being broken up into smaller more manageable chunks."
#316 
#317 
#320 
Cover the first key parts. Please review.

File: schemas/context/experienceevent.example.2.json [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+  },
+  "xdm:placeContext": {
+    "xdm:localTime": "2018-04-20T17:56:23+09:00",
+    "xdm:geo": {
+      "@id": "https://data.adobe.io/entities/geo/tokyo",
+      "xdm:countryCode": "JP",
+      "xdm:stateProvince": "JP-13",
+      "xdm:city": "Tōkyō",
+      "xdm:postalCode": "141-0032",
+      "schema:latitude": 35.6185,
```
Comment (lrosenthol) [inline schemas/context/experienceevent.example.2.json]: is schema:lat & long, the correct names?  doesn't seem so...

**Comment by cdegroot-adobe** [comment]
Completed in other PRs - Closing
